### PR TITLE
Destination Abstraction refactoring with redshift

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
Refactoring destinations v2

### CODE REVIEW

File: airbyte-workers/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java [hunk 1]
```diff
@@ -52,6 +52,7 @@
 
   public enum DestinationType {
     BIGQUERY,
+    JDBC_POSTGRES,
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:55]: is it valuable to include jdbc in normalization at all? can we just skip it?

File: airbyte-workers/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java [hunk 2]
```diff
@@ -37,6 +37,7 @@
   private static final Map<String, DefaultNormalizationRunner.DestinationType> NORMALIZATION_MAPPING =
       ImmutableMap.<String, DefaultNormalizationRunner.DestinationType>builder()
           .put("airbyte/destination-bigquery", DefaultNormalizationRunner.DestinationType.BIGQUERY)
+          .put("airbyte/destination-jdbc", DefaultNormalizationRunner.DestinationType.JDBC_POSTGRES)
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:40]: same

File: airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py [hunk 3]
```diff
@@ -97,6 +100,24 @@ def transform_bigquery(self, config: dict):
 
         return dbt_config
 
+    def transform_jdbc_postgres(self, config: dict):
```
Comment (cgardens) [inline airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py:103]: same as below, what's the value of doing jdbc postgres in normalization at all?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  public SQLNamingResolvable getNamingResolver() {
+    return namingResolver;
+  }
+
+  @Override
+  public DestinationConsumer<AirbyteMessage> write(JsonNode config, ConfiguredAirbyteCatalog catalog) throws Exception {
+    database = getDatabase(config);
+    final Map<String, DestinationWriteContext> writeConfigs =
+        new DestinationConfiguration(getNamingResolver()).getDestinationWriteContext(config, catalog);
+    DestinationConsumerStrategy buffer = new BufferedStreamConsumer(this, catalog);
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:128]: do we need to be passing `this`? i'm especially nervous about passing it into 3 different places. I think this is asking for a weird circular dependency issue.

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        jdbcConfig.get("username").asText(),
+        jdbcConfig.has("password") ? jdbcConfig.get("password").asText() : null,
+        jdbcConfig.get("jdbc_url").asText(),
+        driverClass,
+        dialect);
+  }
+
+  public abstract JsonNode toJdbcConfig(JsonNode config);
+
+  protected String getCurrentDatabaseName(DSLContext ctx) {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:114]: this is protected to handle cases that won't work jooq?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    return ctx.select(currentSchema()).fetch().get(0).getValue(0, String.class);
+  }
+
+  @Override
+  public SQLNamingResolvable getNamingResolver() {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:119]: why is this public? can it be private? i'm only seeing it used on line 127 and then in tests. okay to do `@VisibleForTesting` if it's necessary, but since it is dependency injected, i think it should be needed?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java (hunk 5 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:114]: maybe just leave it private until we fine reason to expose it?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        "CREATE TABLE IF NOT EXISTS %s.%s ( \n"
+            + "ab_id VARCHAR PRIMARY KEY,\n"
+            + "%s JSONB,\n"
+            + "emitted_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP\n"
+            + ");\n",
+        schemaName, tableName, COLUMN_NAME);
+  }
+
+  @Override
+  public String insertBufferedRecordsQuery(int batchSize, CloseableQueue<byte[]> writeBuffer, String schemaName, String tableName) {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java:77]: what does this needed to do this query? assuming for a second that we are only supporting jooq compatible jdbc databases for now, can `AbstractJdbcDestination` handle this?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/InsertTableOperations.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.base;
+
+public interface InsertTableOperations extends TableCreationOperations {
```
Comment (cgardens) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/InsertTableOperations.java:27]: why is this extending `TableCreationOperations`?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/StandardSQLNaming.java [hunk 9]
```diff
@@ -44,6 +44,16 @@ public String getTmpTableName(String streamName) {
     return convertStreamName(streamName + "_" + Instant.now().toEpochMilli());
   }
 
+  @Override
+  public String getSchemaName(String schemaName) {
```
Comment (cgardens) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/StandardSQLNaming.java:48]: why are these needed?

**Review by cgardens — APPROVED** [review]
I think one thing that'll decrease confusion for now is just focusing on doing `JdbcDestination` with the assumption that everything using it is jooq compliant. I think that'll simplify everything alot and require a lot less overriding of methods.

It is pretty hard for me to piece together how the various operations interfaces work, but I think I get the idea though. I think we can go with this and then iterate more on it later. We'll get there!!

File: airbyte-commons/src/main/java/io/airbyte/commons/text/Names.java [hunk 10]
```diff
@@ -43,4 +43,12 @@ public static String toAlphanumericAndUnderscore(String s) {
         .replaceAll(NON_ALPHANUMERIC_AND_UNDERSCORE_PATTERN, "_");
   }
 
+  public static String concatNames(String name, String s) {
```
Comment (michel-tricot) [inline airbyte-commons/src/main/java/io/airbyte/commons/text/Names.java:46]: This function is oddly specific with a non descriptive name, yet in the commons library. 

So, either you call it concatQuotedNames and its behavior is quote(unquote(name) + unquote(s)) or it shouldn't in this file.

can you add a test?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/BufferedWriteOperations.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.base;
+
+import io.airbyte.commons.lang.CloseableQueue;
+
+public interface BufferedWriteOperations {
+
+  void insertBufferedRecords(int batchSize, CloseableQueue<byte[]> writeBuffer, String schemaName, String tableName) throws Exception;
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/BufferedWriteOperations.java:31]: The names are `schemaName` and `tableName` sound very warehouse specific but the iface name isn't. It should since this file lives in a general package.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.text.Names;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.SyncMode;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DestinationConfiguration {
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConfiguration.java:35]: Same problem here with the naming. There is a assumption with SQL but it is not visible in the classname when this file lives in a general package.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/TmpToFinalTable.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.base;
+
+import java.util.Map;
+
+public interface TmpToFinalTable {
+
+  void setContext(Map<String, DestinationCopyContext> configs);
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/TmpToFinalTable.java:31]: The lifecycle of this iface isn't great. If you don't know what you're doing then you might be calling execute without the setContext

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/TruncateInsertIntoConsumer.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  @Override
+  public void setContext(Map<String, DestinationCopyContext> configs) {
+    copyConfigs = configs;
+  }
+
+  @Override
+  public void execute() throws Exception {
+    for (Entry<String, DestinationCopyContext> entry : copyConfigs.entrySet()) {
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/TruncateInsertIntoConsumer.java:48]: given the lifecycle, you should have a Precondition in the execute so that it is "documented" copyConfigs should be set through setContext

**Review by michel-tricot — APPROVED** [review]
The abstractions look a lot better than the previous PR. I still think they can be simplified, we can do it as a later step.

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java (hunk 6 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:119]: it's public because it comes from the interface, if i put private it doesn't compile?

```
'getNamingResolver()' in 'io.airbyte.integrations.destination.jdbc.AbstractJdbcDestination' clashes with 'getNamingResolver()' in 'io.airbyte.integrations.base.Destination'; attempting to assign weaker access privileges ('private'); was 'public'
```

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java (hunk 7 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java:77]: it's tied to the create schema query because it depends on the types of the columns and it affects the implementation of the insert query too...

Thus since JSONB for example is specific to postgres, you asked me to put it outside of AbstractJdbcDestination

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java (hunk 4 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:128]: I don't know what to do, should i make an inner class then?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java (hunk 7 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcDestination.java:77]: ok..

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java (hunk 6 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java:118]: okay. i guess i'm a bit confused about why it's in the destination interface at all. it seems like we only use it internally to destination and don't really intend to use it externally (at least for now). we can figure this out later.

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java (hunk 4 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java]: yeah, that works.

**Review by cgardens — APPROVED** [review]
